python/pyspark/errors/utils.py

```diff
@@ -15,12 +15,22 @@
 # limitations under the License.
 #
 
+import builtins
 import re
-from typing import Dict, Match
+import functools
+import inspect
+import threading
+from typing import Any, Callable, Dict, Match, TypeVar, Type
+
+from IPython import get_ipython
```

Comment:
I don't think `IPython` module is always available.

Reply:
Let me do the Notebook support as follow-up or after completing the initial prototype reviews.

Because maybe I think we need separate discussion about making `IPython` as a project requirements or something?